src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -412,6 +413,15 @@ protected static ComputedInstanceFieldLayout ComputeSequentialFieldLayout(Metada
                 cumulativeInstanceFieldPos = LayoutInt.Max(cumulativeInstanceFieldPos, new LayoutInt(layoutMetadata.Size));
             }
 
+            // ./vm/methodtablebuilder.cpp
+            // This is required because native layout of System.Decimal causes it to be aligned
+            // differently to the layout of the native DECIMAL structure, which will cause
+            // data misalignent exceptions if Decimal is embedded in another type.
+            if (type.IsValueType && type.Context.Target.Architecture == TargetArchitecture.ARM && InteropTypes.IsSystemDecimal(type.Context, type))
```

Comment:
I think it would be easiest to check for CoreLib types by name in a method called from here, and apply the layout quirks that way.

Similar to how `MethodTableBuilder::CheckForSystemTypes` is done. I expect we will need most quirks from this method eventually.

Reply:
Looking at `MethodTableBuilder::CheckForSystemTypes` - will we need anything else besides the Decimal addition? We already treat `Vector<T>` and the other vectors the way I propose for Decimal as well. I don't particularly care about the mechanism, but I do care about consistency.

Reply:
1. Pro: Checking type in `ReadyToRunCompilerContext` allows to don't move `InteropTypes` to `ILC.TypeSystem.R2R` and allows to use existing `InteropTypes.IsDecimal`
2. Pro: Checking type in `ReadyToRunCompilerContext` removes checking of platform via `ReadyToRunCompilerContext` constructor
3. Contra: Checking type in `ReadyToRunCompilerContext` requires to write many code to implement other methods of `FieldLayoutAlgorithm`

4. Pro:  checking type in `MetadataFieldLayoutAlgorithm.cs` exists already and is almost one-liner  
5. Contra: copy-paste of `isDecimal` required or moving of `InteropTypes` to `ILC.TypeSystem.R2R` 
6. Pro: Moving is already done
7. Contra: More changes in a branch  for `static struct` will be to mitigate second half of the 10w5d.cs test.
8. Contra: Colleagues say, more types with broken alignment on ARM32 may arise.

So, Where? Please tell me the consensus!

Reply:
We also faced with similar issues in Nullable types. If there are already exceptions for vectors and other types in `ReadyToRunCompilerContext` I think it would be preferred place.

Reply:
What is the problem with Nullable types? Nullable types should not have any layout quirks.

Reply:
Ups, its nullable long enum:
```C#
public enum LongE : long { start = 1, }
```
from `T/Directed/nullabletypes/castclassenum_ro/castclassenum::NullableTest3()`. Look it's not `Decimal` case.

Reply:
If a nullable long enum isn't handled correctly, that is a bug in the general purpose type layout logic, not something that should be special cased.

Reply:
LongE & so seem are fixed by #32663 / #32664.
But I afraid it is only masking of bug:
When (before) `jit/methodical/fp/exgen/10w5d.cs` `Func_*` don't compiled by `cg2` but delayed to `jit` while `ctor`s is done by cg2.  So some difference in placement of static structs was (IMHO) between `cg2` & `jit` compilers was and leaded to missfunction. 
Right now placement of statics and using of it is by `cg2` and you can't see difference with jit's ideas.

Reply:
For Decimal, I am wondering whether we should fix the Decimal implementation to make this special casing unncessary. I will take a look at what it would take today.

> LongE & so seem are fixed by #32663 / #32664.

I have reverted this workaround a few days ago. Could you please check whether it is still a problem on current master?

Reply:
>> #32663 / #32664.

>I have reverted this workaround a few days ago. Could you please check whether it is still a problem on current master?

With 4e5c8c006122c (today's (20200629) master) the `jit/methodical/fp/exgen/10w5d.cs` is ok.
Colleagues say, `JIT/Directed/nullabletypes/castclassenum_ro/castclassenum_ro.dll` is fixed too.

How to force 'delayed compilation' back I don't know (`eval` does not exists, imho). Sorry.

Reply:
#38603 should be a cleaner fix for the Decimal issue.